File: Tests/HHServicesTests.m
//
//  HHServicesTests.m
//  HHServices Tests
//
//  Tests for HHServices DNS-SD framework
//

#import <XCTest/XCTest.h>
#import "HHService.h"
#import "HHServiceBrowser.h"
#import "HHServicePublisher.h"
#import "HHServiceValidation.h"

@interface HHServicesTests : XCTestCase <HHServiceBrowserDelegate, HHServicePublisherDelegate, HHServiceDelegate>
@property (nonatomic, strong) HHServiceBrowser *browser;
@property (nonatomic, strong) HHServicePublisher *publisher;
@property (nonatomic, strong) XCTestExpectation *browseExpectation;
@property (nonatomic, strong) XCTestExpectation *publishExpectation;
@property (nonatomic, strong) XCTestExpectation *resolveExpectation;
@property (nonatomic, strong) NSMutableArray<HHService *> *discoveredServices;
@end

@implementation HHServicesTests

- (void)setUp {
    [super setUp];
    self.discoveredServices = [NSMutableArray array];
}

- (void)tearDown {
    [self.browser endBrowse];
    [self.publisher endPublish];
    self.browser = nil;
    self.publisher = nil;
    [self.discoveredServices removeAllObjects];
    [super tearDown];
}

#pragma mark - Basic Initialization Tests

- (void)testServiceBrowserInitialization {
    HHServiceBrowser *browser = [[HHServiceBrowser alloc] initWithType:@"_test._tcp." domain:@"local."];
    XCTAssertNotNil(browser, @"Browser should be initialized");
    XCTAssertEqualObjects(browser.type, @"_test._tcp", @"Service type should match");
    XCTAssertEqualObjects(browser.domain, @"local.", @"Domain should match");
}

- (void)testServicePublisherInitialization {
    HHServicePublisher *publisher = [[HHServicePublisher alloc] initWithName:@"TestService" 
                                                                        type:@"_test._tcp." 
                                                                      domain:@"local." 
                                                                     txtData:nil 
                                                                        port:12345];
    XCTAssertNotNil(publisher, @"Publisher should be initialized");
    XCTAssertEqualObjects(publisher.name, @"TestService", @"Service name should match");
    XCTAssertEqualObjects(publisher.type, @"_test._tcp.", @"Service type should match");
    // Note: HHServicePublisher stores port internally but doesn't expose it as a property
}

- (void)testServiceInitialization {
    HHService *service = [[HHService alloc] initWithName:@"TestService" 
                                                    type:@"_test._tcp." 
                                                  domain:@"local."];
    XCTAssertNotNil(service, @"Service should be initialized");
    XCTAssertEqualObjects(service.name, @"TestService", @"Service name should match");
    XCTAssertEqualObjects(service.type, @"_test._tcp.", @"Service type should match");
    XCTAssertEqualObjects(service.domain, @"local.", @"Domain should match");
}

#pragma mark - Publishing and Browsing Tests

- (void)testServicePublishingAndDiscovery {
    // This test requires actual network access and may fail in isolated test environments
    // It's included as an example of integration testing
    
    self.publishExpectation = [self expectationWithDescription:@"Service should be published"];
    self.browseExpectation = [self expectationWithDescription:@"Service should be discovered"];
    
    // Start publishing
    NSData *txtData = [HHServiceValidation txtDataFromDictionary:@{@"test": @"data"}];
    self.publisher = [[HHServicePublisher alloc] initWithName:@"HHServicesTest" 
                                                         type:@"_hhtest._tcp." 
                                                       domain:@"local." 
                                                      txtData:txtData
                                                         port:54321];
    self.publisher.delegate = self;
    BOOL publishStarted = [self.publisher beginPublish];
    XCTAssertTrue(publishStarted, @"Publishing should start successfully");
    
    // Start browsing after a short delay to ensure publishing has started
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        self.browser = [[HHServiceBrowser alloc] initWithType:@"_hhtest._tcp." domain:@"local."];
        self.browser.delegate = self;
        BOOL browseStarted = [self.browser beginBrowse];
        XCTAssertTrue(browseStarted, @"Browsing should start successfully");
    });
    
    // Wait for both operations with timeout
    [self waitForExpectationsWithTimeout:10.0 handler:^(NSError *error) {
        if (error) {
            XCTFail(@"Test timed out: %@", error);
        }
    }];
}

- (void)testBluetoothOnlyBrowsing {
    self.browser = [[HHServiceBrowser alloc] initWithType:@"_hhtest._tcp." domain:@"local."];
    self.browser.delegate = self;
    
    // Test Bluetooth-only browsing
    BOOL browseStarted = [self.browser beginBrowseOverBluetoothOnly];
    // Note: This may fail if Bluetooth is not available or permissions are not granted
    if (!browseStarted) {
        XCTSkip(@"Bluetooth browsing not available in this test environment");
    }
    
    XCTAssertTrue(browseStarted, @"Bluetooth-only browsing should start if available");
    [self.browser endBrowse];
}

#pragma mark - Service Resolution Tests

- (void)testServiceResolution {
    HHService *service = [[HHService alloc] initWithName:@"TestService" 
                                                    type:@"_test._tcp." 
                                                  domain:@"local."];
    service.delegate = self;
    
    // Test different resolution methods
    BOOL resolveStarted = [service beginResolve];
    // Note: This will likely fail without an actual service to resolve
    if (!resolveStarted) {
        XCTSkip(@"Service resolution requires an actual service on the network");
    }
    
    [service endResolve];
}

#pragma mark - HHServiceBrowserDelegate

- (void)serviceBrowser:(HHServiceBrowser *)serviceBrowser didFindService:(HHService *)service moreComing:(BOOL)moreComing {
    [self.discoveredServices addObject:service];
    
    if ([service.name isEqualToString:@"HHServicesTest"] && !moreComing) {
        [self.browseExpectation fulfill];
        
        // Test service resolution
        self.resolveExpectation = [self expectationWithDescription:@"Service should be resolved"];
        service.delegate = self;
        [service beginResolve];
    }
}

- (void)serviceBrowser:(HHServiceBrowser *)serviceBrowser didRemoveService:(HHService *)service moreComing:(BOOL)moreComing {
    [self.discoveredServices removeObject:service];
}

#pragma mark - HHServicePublisherDelegate

- (void)servicePublisherDidPublish:(HHServicePublisher *)servicePublisher {
    [self.publishExpectation fulfill];
}

- (void)serviceDidNotPublish:(HHServicePublisher *)servicePublisher {
    XCTFail(@"Service publishing failed");
    [self.publishExpectation fulfill];
}

#pragma mark - HHServiceDelegate

- (void)serviceDidResolve:(HHService *)service moreComing:(BOOL)moreComing {
    if (!moreComing) {
        XCTAssertNotNil(service.resolvedHostName, @"Service should have resolved host name");
        XCTAssertGreaterThan(service.resolvedPortNumber, 0, @"Service should have valid port");
        [self.resolveExpectation fulfill];
    }
}

- (void)serviceDidNotResolve:(HHService *)service {
    XCTFail(@"Service resolution failed");
    [self.resolveExpectation fulfill];
}

#pragma mark - Error Code Tests

- (void)testPolicyDeniedErrorHandling {
    // Test handling of iOS 14+ Local Network permission denial
    NSError *policyError = [NSError errorWithDomain:@"com.dns_sd" code:-65570 userInfo:nil];
    
    // Verify we can detect policy denied errors
    XCTAssertEqual(policyError.code, -65570, @"Should recognize policy denied error code");
}

#pragma mark - TXT Record Tests

- (void)testTXTRecordHandling {
    NSDictionary *txtDict = @{@"key1": @"value1", @"key2": @"value2"};
    NSData *txtData = [HHServiceValidation txtDataFromDictionary:txtDict];
    XCTAssertNotNil(txtData, @"Should create valid TXT data");
    
    // Validate the TXT data
    XCTAssertTrue([HHServiceValidation isValidTXTRecordData:txtData], @"TXT data should be valid");
    
    // Parse it back
    NSDictionary *parsedDict = [HHServiceValidation dictionaryFromTXTData:txtData];
    XCTAssertEqualObjects(parsedDict[@"key1"], @"value1", @"Should parse key1 correctly");
    XCTAssertEqualObjects(parsedDict[@"key2"], @"value2", @"Should parse key2 correctly");
    
    HHServicePublisher *publisher = [[HHServicePublisher alloc] initWithName:@"TestService" 
                                                                        type:@"_test._tcp." 
                                                                      domain:@"local." 
                                                                     txtData:txtData 
                                                                        port:12345];
    XCTAssertNotNil(publisher.txtData, @"TXT data should be set");
}

#pragma mark - Validation Tests

- (void)testServiceNameValidation {
    // Valid names
    XCTAssertTrue([HHServiceValidation isValidServiceName:@"MyService"], @"Normal name should be valid");
    XCTAssertTrue([HHServiceValidation isValidServiceName:@"Service with spaces"], @"Spaces should be allowed");
    XCTAssertTrue([HHServiceValidation isValidServiceName:@"Émoji-Service"], @"UTF-8 should be allowed");
    
    // Invalid names
    XCTAssertFalse([HHServiceValidation isValidServiceName:nil], @"nil should be invalid");
    XCTAssertFalse([HHServiceValidation isValidServiceName:@""], @"Empty string should be invalid");
    NSString *tooLong = [@"" stringByPaddingToLength:64 withString:@"x" startingAtIndex:0];
    XCTAssertFalse([HHServiceValidation isValidServiceName:tooLong], @"Name over 63 chars should be invalid");
    XCTAssertFalse([HHServiceValidation isValidServiceName:@"Service\nNewline"], @"Control characters should be invalid");
}

- (void)testServiceTypeValidation {
    // Valid types
    XCTAssertTrue([HHServiceValidation isValidServiceType:@"_http._tcp."], @"Standard HTTP type should be valid");
    XCTAssertTrue([HHServiceValidation isValidServiceType:@"_test._udp."], @"UDP type should be valid");
    XCTAssertTrue([HHServiceValidation isValidServiceType:@"_my-service._tcp."], @"Hyphen should be allowed");
    
    // Invalid types
    XCTAssertFalse([HHServiceValidation isValidServiceType:nil], @"nil should be invalid");
    XCTAssertFalse([HHServiceValidation isValidServiceType:@"http._tcp."], @"Missing underscore should be invalid");
    XCTAssertFalse([HHServiceValidation isValidServiceType:@"_http.tcp."], @"Missing underscore before protocol should be invalid");
    XCTAssertFalse([HHServiceValidation isValidServiceType:@"_http._xyz."], @"Invalid protocol should be invalid");
    XCTAssertFalse([HHServiceValidation isValidServiceType:@"_toolongservicename._tcp."], @"Service name over 15 chars should be invalid");
}

- (void)testDomainValidation {
    // Valid domains
    XCTAssertTrue([HHServiceValidation isValidDomain:@"local."], @"local. should be valid");
    XCTAssertTrue([HHServiceValidation isValidDomain:@"example.com."], @"example.com. should be valid");
    
    // Invalid domains
    XCTAssertFalse([HHServiceValidation isValidDomain:nil], @"nil should be invalid");
    XCTAssertFalse([HHServiceValidation isValidDomain:@""], @"Empty string should be invalid");
    XCTAssertFalse([HHServiceValidation isValidDomain:@"local"], @"Missing dot should be invalid");
    XCTAssertFalse([HHServiceValidation isValidDomain:@"-invalid.com."], @"Starting with hyphen should be invalid");
}

- (void)testServiceNameSanitization {
    XCTAssertEqualObjects([HHServiceValidation sanitizedServiceName:@"Valid Name"], @"Valid Name");
    XCTAssertEqualObjects([HHServiceValidation sanitizedServiceName:@"Name\nWith\tControl"], @"NameWithControl");
    NSString *tooLong = [@"" stringByPaddingToLength:100 withString:@"x" startingAtIndex:0];
    XCTAssertEqual([HHServiceValidation sanitizedServiceName:tooLong].length, 63);
    XCTAssertEqualObjects([HHServiceValidation sanitizedServiceName:@""], @"Service");
    XCTAssertEqualObjects([HHServiceValidation sanitizedServiceName:nil], @"Service");
}

@end

File: Tests/HHServicesTests.swift
import XCTest
@testable import HHServices

final class HHServicesTests: XCTestCase {
    
    // MARK: - Service Validation Tests
    
    func testServiceNameValidation() {
        // Valid names
        XCTAssertTrue(ServiceValidation.isValidServiceName("MyService"))
        XCTAssertTrue(ServiceValidation.isValidServiceName("Service-123"))
        XCTAssertTrue(ServiceValidation.isValidServiceName("A"))
        XCTAssertTrue(ServiceValidation.isValidServiceName("My Printer @ Home"))
        XCTAssertTrue(ServiceValidation.isValidServiceName("Service with émoji 🎉")) // Unicode is allowed
        
        // Invalid names
        XCTAssertFalse(ServiceValidation.isValidServiceName(""))
        XCTAssertFalse(ServiceValidation.isValidServiceName(String(repeating: "a", count: 100))) // Too long
        
        // Edge cases
        let exactly63Bytes = String(repeating: "a", count: 63)
        XCTAssertTrue(ServiceValidation.isValidServiceName(exactly63Bytes))
        
        let exactly64Bytes = String(repeating: "a", count: 64)
        XCTAssertFalse(ServiceValidation.isValidServiceName(exactly64Bytes))
    }
    
    func testServiceTypeValidation() {
        // Valid types
        XCTAssertTrue(ServiceValidation.isValidServiceType("_http._tcp."))
        XCTAssertTrue(ServiceValidation.isValidServiceType("_http._tcp"))
        XCTAssertTrue(ServiceValidation.isValidServiceType("_printer._tcp."))
        XCTAssertTrue(ServiceValidation.isValidServiceType("_test._udp."))
        XCTAssertTrue(ServiceValidation.isValidServiceType("_my-service._tcp."))
        
        // Invalid types
        XCTAssertFalse(ServiceValidation.isValidServiceType(""))
        XCTAssertFalse(ServiceValidation.isValidServiceType("http._tcp.")) // Missing leading underscore
        XCTAssertFalse(ServiceValidation.isValidServiceType("_http.tcp.")) // Missing underscore before tcp
        XCTAssertFalse(ServiceValidation.isValidServiceType("_http._xyz.")) // Invalid protocol
        XCTAssertFalse(ServiceValidation.isValidServiceType("_._tcp.")) // Empty service name
    }
    
    func testDomainValidation() {
        // Valid domains
        XCTAssertTrue(ServiceValidation.isValidDomain("local."))
        XCTAssertTrue(ServiceValidation.isValidDomain("local"))
        XCTAssertTrue(ServiceValidation.isValidDomain(""))  // Empty is valid (means default)
        XCTAssertTrue(ServiceValidation.isValidDomain("example.com."))
        XCTAssertTrue(ServiceValidation.isValidDomain("sub.example.com."))
        
        // Invalid domains
        XCTAssertFalse(ServiceValidation.isValidDomain("-invalid.com")) // Starts with hyphen
        XCTAssertFalse(ServiceValidation.isValidDomain("invalid-.com")) // Ends with hyphen
        XCTAssertFalse(ServiceValidation.isValidDomain("invalid..com")) // Double dot
        
        // Edge cases
        let longLabel = String(repeating: "a", count: 63)
        XCTAssertTrue(ServiceValidation.isValidDomain("\(longLabel).com"))
        
        let tooLongLabel = String(repeating: "a", count: 64)
        XCTAssertFalse(ServiceValidation.isValidDomain("\(tooLongLabel).com"))
    }
    
    func testServiceNameSanitization() {
        // Test control character removal
        let nameWithControl = "Service\u{0000}Name\u{007F}"
        XCTAssertEqual(ServiceValidation.sanitizeServiceName(nameWithControl), "ServiceName")
        
        // Test truncation
        let longName = String(repeating: "a", count: 100)
        let sanitized = ServiceValidation.sanitizeServiceName(longName)
        XCTAssertTrue(sanitized.data(using: .utf8)!.count <= 63)
        
        // Test empty becomes default
        XCTAssertEqual(ServiceValidation.sanitizeServiceName(""), "Service")
        XCTAssertEqual(ServiceValidation.sanitizeServiceName("\u{0000}"), "Service")
    }
    
    // MARK: - TXT Record Tests
    
    func testTXTRecordParsing() {
        // Create a TXT record with key-value pairs
        let dict = [
            "version": "1.0",
            "path": "/printer",
            "paper": "A4"
        ]
        
        let txtRecord = TXTRecord(dictionary: dict)
        
        // Test retrieval
        XCTAssertEqual(txtRecord["version"], "1.0")
        XCTAssertEqual(txtRecord["path"], "/printer")
        XCTAssertEqual(txtRecord["paper"], "A4")
        XCTAssertNil(txtRecord["nonexistent"])
        
        // Test keys
        XCTAssertTrue(txtRecord.hasKey("version"))
        XCTAssertFalse(txtRecord.hasKey("nonexistent"))
    }
    
    func testTXTRecordWithEmptyValues() {
        let dict = [
            "flag": "",
            "key": "value"
        ]
        
        let txtRecord = TXTRecord(dictionary: dict)
        XCTAssertEqual(txtRecord["flag"], "")
        XCTAssertEqual(txtRecord["key"], "value")
    }
    
    // MARK: - Socket Address Tests
    
    func testSocketAddressIPv4() {
        // Create IPv4 address data
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = UInt16(8080).bigEndian
        addr.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        
        let data = Data(bytes: &addr, count: Int(addr.sin_len))
        let socketAddr = SocketAddress(data: data)
        
        XCTAssertEqual(socketAddr.family, .ipv4)
        XCTAssertEqual(socketAddr.port, 8080)
    }
    
    func testSocketAddressIPv6() {
        // Create IPv6 address data
        var addr = sockaddr_in6()
        addr.sin6_family = sa_family_t(AF_INET6)
        addr.sin6_port = UInt16(9090).bigEndian
        addr.sin6_len = UInt8(MemoryLayout<sockaddr_in6>.size)
        
        let data = Data(bytes: &addr, count: Int(addr.sin6_len))
        let socketAddr = SocketAddress(data: data)
        
        XCTAssertEqual(socketAddr.family, .ipv6)
        XCTAssertEqual(socketAddr.port, 9090)
    }
    
    // MARK: - Service Tests
    
    func testServiceCreation() {
        let service = Service(
            name: "TestService",
            type: "_http._tcp.",
            domain: "local.",
            interfaceIndex: 0
        )
        
        XCTAssertEqual(service.name, "TestService")
        XCTAssertEqual(service.type, "_http._tcp.")
        XCTAssertEqual(service.domain, "local.")
        XCTAssertFalse(service.isResolved)
        XCTAssertFalse(service.hasAddresses)
    }
    
    func testServiceEquality() {
        let service1 = Service(name: "Test", type: "_http._tcp.", domain: "local.")
        let service2 = Service(name: "Test", type: "_http._tcp.", domain: "local.")
        let service3 = Service(name: "Other", type: "_http._tcp.", domain: "local.")
        
        XCTAssertEqual(service1, service2)
        XCTAssertNotEqual(service1, service3)
    }
    
    func testServiceResolution() {
        let service = Service(name: "Test", type: "_http._tcp.", domain: "local.")
        
        // Simulate resolution
        service.updateResolvedInfo(
            hostName: "test.local.",
            port: 8080,
            txtData: nil
        )
        
        XCTAssertTrue(service.isResolved)
        XCTAssertEqual(service.hostName, "test.local.")
        XCTAssertEqual(service.port, 8080)
    }
    
    // MARK: - Error Tests
    
    func testErrorDescriptions() {
        let error1 = HHServicesError.invalidServiceType
        XCTAssertNotNil(error1.errorDescription)
        
        let error2 = HHServicesError.localNetworkPermissionDenied
        XCTAssertNotNil(error2.errorDescription)
        XCTAssertNotNil(error2.failureReason)
        
        let error3 = HHServicesError.dnsFailed(-65563)
        XCTAssertNotNil(error3.errorDescription)
    }
    
    // MARK: - Integration Tests
    
    func testServiceBrowserCreation() {
        let browser = ServiceBrowser(type: "_http._tcp.", domain: "local.")
        
        XCTAssertEqual(browser.serviceType, "_http._tcp.")
        XCTAssertEqual(browser.domain, "local.")
        XCTAssertEqual(browser.services.count, 0)
    }
    
    func testServicePublisherCreation() {
        let publisher = ServicePublisher(
            name: "TestService",
            type: "_http._tcp.",
            port: 8080
        )
        
        XCTAssertEqual(publisher.name, "TestService")
        XCTAssertEqual(publisher.type, "_http._tcp.")
        XCTAssertEqual(publisher.port, 8080)
    }
    
    func testInvalidServiceTypeThrows() {
        let browser = ServiceBrowser(type: "invalid", domain: "local.")
        
        XCTAssertThrowsError(try browser.startBrowsing()) { error in
            XCTAssertTrue(error is HHServicesError)
            if let hhError = error as? HHServicesError {
                XCTAssertEqual(hhError, HHServicesError.invalidServiceType)
            }
        }
    }
    
    func testInvalidDomainThrows() {
        let browser = ServiceBrowser(type: "_http._tcp.", domain: "-invalid")
        
        XCTAssertThrowsError(try browser.startBrowsing()) { error in
            XCTAssertTrue(error is HHServicesError)
            if let hhError = error as? HHServicesError {
                XCTAssertEqual(hhError, HHServicesError.invalidDomain)
            }
        }
    }
    
    // MARK: - Async Tests
    
    @available(iOS 13.0, *)
    func testAsyncBrowsingInterface() async throws {
        // Test that the async interface exists and compiles
        // We don't actually browse to avoid network dependencies in tests
        let browser = ServiceBrowser(type: "_test._tcp.", domain: "local.")
        
        // Just verify we can create the browser and it has the expected properties
        XCTAssertEqual(browser.serviceType, "_test._tcp.")
        XCTAssertEqual(browser.domain, "local.")
        XCTAssertEqual(browser.services.count, 0)
        
        // Verify stop doesn't crash
        browser.stop()
        
        // The test passes if the interface exists
        XCTAssertTrue(true)
    }
    
    @available(iOS 13.0, *)
    func testServiceResolverInterface() async throws {
        // Test that the resolver interface exists
        let service = Service(name: "Test", type: "_test._tcp.", domain: "local.")
        let resolver = ServiceResolver(service: service, timeout: 0.1)
        
        // Stop immediately to avoid network calls
        resolver.stop()
        
        // The test passes if we can create and stop the resolver
        XCTAssertTrue(true)
    }
}